# Lab | Pandas

Objective: practice how to use the pandas library in Python for data analysis and manipulation.

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [2]:
#import pandas library 
import pandas as pd
df_original = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")
df_original

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [3]:
# create a copy of the original file
df_copy = df_original.copy()

In [4]:
# for the original DataFrame print the number of rows and columns 
num_rows, num_columns = df_original.shape
print("Number of rows:", num_rows)
print("Number of columns:", num_columns)

Number of rows: 4008
Number of columns: 11


In [15]:
# retrieves the data types of the columns in the DataFrame
data_types = df_original.dtypes
for column, dtype in data_types.items(): #iterates over each column and its corresponding data type.
    print(f"Column: {column}") # For each column, print the column name and data type.
    print(f"Data Type: {dtype}")
    print()

Column: Customer
Data Type: object

Column: ST
Data Type: object

Column: GENDER
Data Type: object

Column: Education
Data Type: object

Column: Customer Lifetime Value
Data Type: object

Column: Income
Data Type: float64

Column: Monthly Premium Auto
Data Type: float64

Column: Number of Open Complaints
Data Type: object

Column: Policy Type
Data Type: object

Column: Vehicle Class
Data Type: object

Column: Total Claim Amount
Data Type: float64



Column:Number of Open Complaints
Data Type: object
To ---> float 

Column:Customer Lifetime Value
Data Type: object
To ---> float 

In [19]:
#calculates the number of unique values in each column of the DataFrame
unique_counts = df_original.nunique()
categorical_columns = unique_counts[unique_counts <= 10].index.tolist() #for each column calculate the number of unique values and check if the value is <= than the number of rows  
print("Number of unique values for each column:")
print(unique_counts)
print("\nCategorical columns:")
print(categorical_columns)

Number of unique values for each column:
Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

Categorical columns:
['ST', 'GENDER', 'Education', 'Number of Open Complaints', 'Policy Type', 'Vehicle Class']


**Based on the provided information:**
The insights extract from our DataFrame provide us an understanding of the diversity and range of values present in each column.
We can regard thar the dataset consists of records from 1071 unique customers.

And for each column we have the unique values for each row, for example:
**ST:** This column represents the state or region where the customers are located and there are 8 different regions on this dataset

**GENDER:** This column indicate the different gender categories for customers and there are 5 unique values in this columns

**Education:** this column represents the differents levels of education of the customers and there are 6 unique values in each column

(...) and so on.

At the end we can understand that our DataFrame has 6 columns that are categorical, what means that these columns contain qualitative variables with distinct categories or levels.

In [7]:
numerical_columns = df_original.select_dtypes(include=[int, float]).columns
print(f"Numerical Columns: {numerical_columns}")

Numerical Columns: Index(['Income', 'Monthly Premium Auto', 'Total Claim Amount'], dtype='object')


In [8]:
summary_stats = round(df_original[numerical_columns].describe(),2) # round function to 2 decimals
print(summary_stats)

         Income  Monthly Premium Auto  Total Claim Amount
count   1071.00               1071.00             1071.00
mean   39295.70                193.23              404.99
std    30469.43               1601.19              293.03
min        0.00                 61.00                0.38
25%    14072.00                 68.00              202.16
50%    36234.00                 83.00              354.73
75%    64631.00                109.50              532.80
max    99960.00              35354.00             2893.24


In [9]:
print(categorical_columns)

['ST', 'GENDER', 'Education', 'Number of Open Complaints', 'Policy Type', 'Vehicle Class']


In [10]:
summary_stats_cat = df_original[categorical_columns].describe() 
print(summary_stats_cat)

            ST GENDER Education Number of Open Complaints    Policy Type  \
count     1071    954      1071                      1071           1071   
unique       8      5         6                         6              3   
top     Oregon      F  Bachelor                    1/0/00  Personal Auto   
freq       320    457       324                       830            780   

        Vehicle Class  
count            1071  
unique              6  
top     Four-Door Car  
freq              576  


## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [11]:
location_series = df_original['ST'].value_counts()
sorted_locations = location_series.sort_values()
top_5_less_common = sorted_locations.head(5)
print(top_5_less_common)

AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: ST, dtype: int64


### Exercise 2


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [16]:
threshold = df_original['Total Claim Amount'].quantile(0.75)
high_claim_customers = df_original[df_original['Total Claim Amount'] > threshold] 
summary_stats = high_claim_customers['Total Claim Amount'].describe().round(2)
print("Summary statistics about the high policy claim amount data")
print(" ")
print(summary_stats)

Summary statistics about the high policy claim amount data
 
count     264.00
mean      782.23
std       292.75
min       537.60
25%       606.52
50%       679.60
75%       851.40
max      2893.24
Name: Total Claim Amount, dtype: float64


### Exercise 3

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [17]:
policy_counts = df_original['Policy Type'].value_counts()
policy_with_highest_count = policy_counts.idxmax()
print("Policy type with the highest number of policies sold:", policy_with_highest_count)

Policy type with the highest number of policies sold: Personal Auto


### Exercise 4

The sales team wants to know if customers with Personal Auto have a income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [18]:
# Create a DataFrame for Personal Auto policies
personal_auto_df = df_original.loc[df_original['Policy Type'] == 'Personal Auto']

# Create a DataFrame for Corporate Auto policies
corporate_auto_df = df_original.loc[df_original['Policy Type'] == 'Corporate Auto']

# Calculate the average income for each policy type
average_income_personal_auto = round(personal_auto_df['Income'].mean(), 2)
average_income_corporate_auto = round(corporate_auto_df['Income'].mean(), 2)

# Print the results
print("Average Income for Personal Auto policies: $", average_income_personal_auto)
print("Average Income for Corporate Auto policies: $", average_income_corporate_auto)

Average Income for Personal Auto policies: $ 38180.7
Average Income for Corporate Auto policies: $ 41390.31
